haystack/preview/components/routers/document_joiner.py

```diff
@@ -0,0 +1,142 @@
+import itertools
+import logging
+from collections import defaultdict
+from math import inf
+from typing import List, Optional
+from canals.component.types import Variadic
+
+from haystack.preview import component, Document
+
+
+logger = logging.getLogger(__name__)
+
+
+@component
+class DocumentJoiner:
+    """
+    A component that joins input lists of documents from multiple connections and outputs them as one list.
+
+    The component allows multiple join modes:
+    * concatenate: combine the documents from multiple nodes. Any duplicate documents are discarded.
+                   The score is only determined by the last node that outputs the document.
+    * merge: merge scores of documents from multiple nodes. Optionally, each input score can be given a different
+             `weight` & a `top_k` limit can be set. This mode can also be used for "reranking" retrieved documents.
+    * reciprocal_rank_fusion: combines the documents based on their rank in multiple nodes.
```

Comment:
```suggestion
    * reciprocal_rank_fusion: combines the Documents into a single list based on their rank in multiple nodes.
```
rank or score? let's stick to one term here, we used score before so let's say: based on the score they got from previous components.